cmd/common/registry.go

```diff
@@ -0,0 +1,377 @@
+			if err != nil {
+				return fmt.Errorf("initializing operators: %w", err)
+			}
+
+			fe := console.NewFrontend()
+			defer fe.Close()
+
+			ctx := fe.GetContext()
+
+			// Handle timeout parameter by adding a timeout to the context
+			if timeout != 0 {
+				tmpCtx, cancel := context.WithTimeout(ctx, time.Duration(timeout)*time.Second)
+				defer cancel()
+				ctx = tmpCtx
+			}
+
+			log := logger.DefaultLogger()
+
+			gadgetCtx := gadgetcontext.New(
+				ctx,
+				"",
+				runtime,
+				gadgetDesc,
+				gadgetParams,
+				operatorsParamCollection,
+				parser,
+				log,
+			)
+
+			if parser == nil {
+				// This kind of gadgets return directly the result instead of
+				// using the parser
+				result, err := runtime.RunGadget(gadgetCtx)
+				if err != nil {
+					return fmt.Errorf("running gadget: %w", err)
+				}
+
+				switch outputMode {
+				default:
+					transformer, ok := gadgetDesc.(gadgets.GadgetOutputFormats)
+					if !ok {
+						return fmt.Errorf("gadget does not provide output formats")
+					}
+					formats, defaultFormat := transformer.OutputFormats()
+					transformed, err := formats[defaultFormat].Transform(result)
+					if err != nil {
+						return fmt.Errorf("transforming gadget result: %w", err)
+					}
+					fe.Output(string(transformed))
+				case utils.OutputModeJSON:
+					fe.Output(string(result))
+				}
+
+				return nil
+			}
+
+			// Add some custom params like filters that are available when using parser
+			if len(filters) > 0 {
+				err = parser.SetFilters(filters)
+				if err != nil {
+					return err // TODO: Wrap
+				}
+				gadgetParams.AddKeyValuePair("columns_filters", strings.Join(filters, ",")) // TODO: maybe encode?! difficult for CRs though
+			}
+
+			if gadgetDesc.Type().CanSort() {
+				sortBy := gadgetParams.Get(gadgets.ParamSortBy).AsStringSlice()
+				err := parser.SetSorting(sortBy)
+				if err != nil {
+					return err // TODO: Wrap
+				}
+			}
+
+			formatter := parser.GetTextColumnsFormatter()
+			formatter.SetShowColumns(showColumns)
+			parser.SetErrorCallback(fe.Error)
+
+			// Wire up callbacks before handing over to runtime depending on the output mode
+			switch outputMode {
+			default:
+				formatter.SetEventCallback(fe.Output)
+
+				// Enable additional output, if the gadget supports it (e.g. profile/cpu)
+				//  TODO: This can be optimized later on
+				formatter.SetEnableExtraLines(true)
+
+				parser.SetEventCallback(formatter.EventHandlerFunc())
+				if gadgetDesc.Type().IsPeriodic() {
+					// In case of periodic outputting gadgets, this is done as full table output, and we need to
+					// clear the screen for every interval, that's why we add fe.Clear here
+					parser.SetEventCallback(formatter.EventHandlerFuncArray(
+						fe.Clear,
+						func() {
+							fe.Output(formatter.FormatHeader())
+						},
+					))
+					break
+				}
+				fe.Output(formatter.FormatHeader())
+				parser.SetEventCallback(formatter.EventHandlerFuncArray())
```

Comment:
Perhaps something like the following is more clear?

```suggestion
				if gadget.Type().IsPeriodic() {
					// In case of periodic outputting gadgets, this is done as full table output, and we need to
					// clear the screen for every interval, that's why we add fe.Clear here
					parser.SetEventCallback(formatter.EventHandlerFuncArray(
						fe.Clear,
						func() {
							fe.Output(formatter.FormatHeader())
						},
					))
				} else {
					parser.SetEventCallback(formatter.EventHandlerFunc())
				}
				fe.Output(formatter.FormatHeader())
				parser.SetEventCallback(formatter.EventHandlerFuncArray())
```

Also notice that for top gadget we'd like to print the headers as soon as possible, for instance in cases like `sudo ./local-gadget top file --interval 5` we don't want the user to wait 5 seconds to get at least the headers printed, otherwise it could seem that the gadget is frozen.

Reply:
We need an `fe.Clear()` before the `fe.Output(formatter.FormatHeader())` and we shouldn't set the array callback for the `IsPeriodic()` case, but I've taken care of the initial printing of the headers now.